reasoner_util/__init__.py

```diff
@@ -156,3 +157,29 @@ def apply_ids(id_map: dict, message_dict: dict) -> None:
         for rnode in result["node_bindings"].values():
             for entry in rnode:
                 entry["id"] = id_map[entry["id"]]
+
+
+def merge_qedges(qedges1: dict, qedges2: dict) -> dict:
+    """Merge qedges: the keys must be the same and the values must be the same.
+    If a key is unique to one edges dict, then the edge will be concatenated to
+    the new edges dictionary. If a particular key exists in both messages but
+    the values do not match, this will result in an error. """
+    qedges1 = copy.deepcopy(qedges1)
+    new_edges = {}
```

Comment:
Consider the following:
```python
new_edges = copy.deepcopy(qedges1)
for qedge_key, qedge_value in qedges2.items():
    if qedge_key in new_edges and qedge_value != new_edges[qedge_key]:
        raise ValueError(
           "Key exists in both messages but values do not match."
        )
    new_edges[qedge_key] = copy.deepcopy(qedge_value)
```
Is this equivalent?

Reply:
it has an equivalent outcome, but we we will be making another deepcopy of qedge pairs even when they already exist in new_edges. is that a problem in terms of memory?

Reply:
Ah, good catch.
```python
new_edges = copy.deepcopy(qedges1)
for qedge_key, qedge_value in qedges2.items():
    if qedge_key not in new_edges:
        new_edges[qedge_key] = copy.deepcopy(qedge_value)
    elif qedge_value != new_edges[qedge_key]:
        raise ValueError(
           "Key exists in both messages but values do not match."
        )
```
Better?

Reply:
Yes, this looks good!